From 8c09c0da5cefa15a8453990f496bc7cd228be6b3 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Fri, 22 Aug 2014 22:43:19 -0400
Subject: [fs] pipe: skip file_update_time on frozen fs

Message-id: <53F7C787.3080001@redhat.com>
Patchwork-id: 88714
O-Subject: [PATCH RHEL7] fs/pipe.c: skip file_update_time on frozen fs
Bugzilla: 1107499
RH-Acked-by: Brian Foster <bfoster@redhat.com>

This is for
Bug 1107499 - fs: writing to full named pipe blocks freezing - rhel7

It's a straightforward backport of this upstream commit in v3.14,
tested with the reproducer in the bug.  This has already been fixed
in RHEL6.

    commit 7e775f46a125f894a1d71e96797c776dbec161f0
    Author: Dmitry Monakhov <dmonakhov@openvz.org>
    Date:   Thu Jan 23 15:55:21 2014 -0800

    fs/pipe.c: skip file_update_time on frozen fs

    Pipe has no data associated with fs so it is not good idea to block
    pipe_write() if FS is frozen, but we can not update file's time on such
    filesystem.  Let's use same idea as we use in touch_time().

    Addresses https://bugzilla.kernel.org/show_bug.cgi?id=65701

    Signed-off-by: Dmitry Monakhov <dmonakhov@openvz.org>
    Reviewed-by: Jan Kara <jack@suse.cz>
    Cc: Al Viro <viro@zeniv.linux.org.uk>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Thanks,
-Eric
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/pipe.c b/fs/pipe.c
index 0e0752e..78fd0d0 100644
--- a/fs/pipe.c
+++ b/fs/pipe.c
@@ -663,10 +663,11 @@ out:
   wake_up_interruptible_sync_poll(&pipe->wait, POLLIN | POLLRDNORM);
   kill_fasync(&pipe->fasync_readers, SIGIO, POLL_IN);
  }
- if (ret > 0) {
+ if (ret > 0 && sb_start_write_trylock(file_inode(filp)->i_sb)) {
   int err = file_update_time(filp);
   if (err)
    ret = err;
+  sb_end_write(file_inode(filp)->i_sb);
  }
  return ret;
 }
-- 
1.7.1